File: python/agentshield/mcp/scan.py
"""MCP tool scanning — description poisoning, typosquat, shadow tools.

Heuristics are deterministic and local (no network, no LLM judge).
"""

from __future__ import annotations

import re
from dataclasses import asdict, dataclass, field
from difflib import SequenceMatcher
from typing import Any, Iterable, Mapping, Sequence


POISON_PATTERNS: list[tuple[str, re.Pattern[str], str, str]] = [
    (
        "instruction_override",
        re.compile(
            r"(ignore|disregard|override).{0,40}(previous|prior|system).{0,20}(instruction|prompt|rule)",
            re.I,
        ),
        "Critical",
        "MCP01",
    ),
    (
        "hidden_exfil",
        re.compile(
            r"(send|exfiltrat|upload|post).{0,40}(secret|api.?key|token|password|credential)",
            re.I,
        ),
        "Critical",
        "MCP01",
    ),
    (
        "privilege_claim",
        re.compile(
            r"(you (must|should) |always )?(grant|enable|use).{0,30}(admin|root|sudo|unrestricted)",
            re.I,
        ),
        "High",
        "MCP01",
    ),
    (
        "do_not_tell_user",
        re.compile(r"(do not|don't|never).{0,20}(tell|inform|reveal|mention).{0,20}(user|human)", re.I),
        "High",
        "MCP01",
    ),
    (
        "tool_priority_hijack",
        re.compile(r"(prefer|always use|must use).{0,30}(this tool|me first)", re.I),
        "Medium",
        "MCP01",
    ),
]


@dataclass
class ScanFinding:
    tool_name: str
    rule_id: str
    severity: str
    taxonomy: str
    message: str
    evidence: str = ""

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


@dataclass
class ScanReport:
    findings: list[ScanFinding] = field(default_factory=list)
    tools_scanned: int = 0

    @property
    def ok(self) -> bool:
        return not any(f.severity in ("Critical", "High") for f in self.findings)

    def to_dict(self) -> dict[str, Any]:
        return {
            "ok": self.ok,
            "tools_scanned": self.tools_scanned,
            "findings": [f.to_dict() for f in self.findings],
        }


def _lookalike_ratio(a: str, b: str) -> float:
    return SequenceMatcher(None, a.lower(), b.lower()).ratio()


def _homoglyph_normalize(name: str) -> str:
    table = str.maketrans(
        {
            "0": "o",
            "1": "l",
            "3": "e",
            "4": "a",
            "5": "s",
            "7": "t",
            "@": "a",
            "$": "s",
        }
    )
    return name.lower().translate(table)


def scan_description(tool: Mapping[str, Any]) -> list[ScanFinding]:
    name = str(tool.get("name", "") or "<unnamed>")
    description = str(tool.get("description", "") or "")
    findings: list[ScanFinding] = []
    for rule_id, pattern, severity, taxonomy in POISON_PATTERNS:
        match = pattern.search(description)
        if match:
            findings.append(
                ScanFinding(
                    tool_name=name,
                    rule_id=rule_id,
                    severity=severity,
                    taxonomy=taxonomy,
                    message=f"Description poisoning heuristic matched: {rule_id}",
                    evidence=match.group(0)[:120],
                )
            )
    return findings


def scan_shadow_and_typosquat(
    tools: Sequence[Mapping[str, Any]],
    *,
    trusted_names: Iterable[str] | None = None,
    similarity_threshold: float = 0.82,
) -> list[ScanFinding]:
    findings: list[ScanFinding] = []
    names = [str(t.get("name", "")).strip() for t in tools if t.get("name")]
    trusted = {n.lower() for n in (trusted_names or [])}

    # Exact duplicate / shadow names
    seen: dict[str, int] = {}
    for n in names:
        key = n.lower()
        seen[key] = seen.get(key, 0) + 1
    for key, count in seen.items():
        if count > 1:
            findings.append(
                ScanFinding(
                    tool_name=key,
                    rule_id="shadow_tool_duplicate",
                    severity="High",
                    taxonomy="MCP03",
                    message=f"Shadow tool: name appears {count} times",
                    evidence=key,
                )
            )

    # Pairwise lookalikes
    for i, a in enumerate(names):
        for b in names[i + 1 :]:
            if a.lower() == b.lower():
                continue
            ratio = _lookalike_ratio(a, b)
            norm_ratio = _lookalike_ratio(_homoglyph_normalize(a), _homoglyph_normalize(b))
            score = max(ratio, norm_ratio)
            if score >= similarity_threshold:
                findings.append(
                    ScanFinding(
                        tool_name=a,
                        rule_id="typosquat_lookalike",
                        severity="High",
                        taxonomy="MCP08",
                        message=f"Typosquat / lookalike pair: {a!r} ~ {b!r} ({score:.2f})",
                        evidence=f"{a}|{b}|{score:.2f}",
                    )
                )

    # Untrusted names when an allowlist of trusted names is provided
    if trusted:
        for n in names:
            if n.lower() not in trusted:
                # only flag if lookalike to a trusted name
                for t in trusted:
                    if _lookalike_ratio(n, t) >= similarity_threshold and n.lower() != t:
                        findings.append(
                            ScanFinding(
                                tool_name=n,
                                rule_id="untrusted_lookalike",
                                severity="High",
                                taxonomy="MCP08",
                                message=f"Untrusted tool looks like trusted {t!r}",
                                evidence=n,
                            )
                        )
                        break

    return findings


def scan_tools(
    tools: Sequence[Mapping[str, Any]],
    *,
    trusted_names: Iterable[str] | None = None,
    similarity_threshold: float = 0.82,
) -> ScanReport:
    """Scan a set of MCP / function tools for poisoning and identity issues."""
    findings: list[ScanFinding] = []
    for tool in tools:
        findings.extend(scan_description(tool))
    findings.extend(
        scan_shadow_and_typosquat(
            tools,
            trusted_names=trusted_names,
            similarity_threshold=similarity_threshold,
        )
    )
    return ScanReport(findings=findings, tools_scanned=len(tools))

File: server/services/mcp/scan.js
'use strict';

const POISON_PATTERNS = [
  {
    rule_id: 'instruction_override',
    severity: 'Critical',
    taxonomy: 'MCP01',
    re: /(ignore|disregard|override).{0,40}(previous|prior|system).{0,20}(instruction|prompt|rule)/i
  },
  {
    rule_id: 'hidden_exfil',
    severity: 'Critical',
    taxonomy: 'MCP01',
    re: /(send|exfiltrat|upload|post).{0,40}(secret|api.?key|token|password|credential)/i
  },
  {
    rule_id: 'privilege_claim',
    severity: 'High',
    taxonomy: 'MCP01',
    re: /(you (must|should) |always )?(grant|enable|use).{0,30}(admin|root|sudo|unrestricted)/i
  },
  {
    rule_id: 'do_not_tell_user',
    severity: 'High',
    taxonomy: 'MCP01',
    re: /(do not|don't|never).{0,20}(tell|inform|reveal|mention).{0,20}(user|human)/i
  },
  {
    rule_id: 'tool_priority_hijack',
    severity: 'Medium',
    taxonomy: 'MCP01',
    re: /(prefer|always use|must use).{0,30}(this tool|me first)/i
  }
];

function ratio(a, b) {
  const s = String(a).toLowerCase();
  const t = String(b).toLowerCase();
  if (s === t) return 1;
  const longer = s.length > t.length ? s : t;
  const shorter = s.length > t.length ? t : s;
  if (!longer.length) return 1;
  let matches = 0;
  for (let i = 0; i < shorter.length; i += 1) {
    if (longer.includes(shorter[i])) matches += 1;
  }
  // Use Levenshtein-lite via shared prefix/suffix + length
  let dp = Array.from({ length: s.length + 1 }, (_, i) => i);
  for (let j = 1; j <= t.length; j += 1) {
    let prev = dp[0];
    dp[0] = j;
    for (let i = 1; i <= s.length; i += 1) {
      const tmp = dp[i];
      dp[i] = s[i - 1] === t[j - 1] ? prev : 1 + Math.min(prev, dp[i], dp[i - 1]);
      prev = tmp;
    }
  }
  const dist = dp[s.length];
  return 1 - dist / Math.max(s.length, t.length, 1);
}

function scanDescription(tool) {
  const name = String(tool.name || '<unnamed>');
  const description = String(tool.description || '');
  const findings = [];
  for (const p of POISON_PATTERNS) {
    const m = description.match(p.re);
    if (m) {
      findings.push({
        tool_name: name,
        rule_id: p.rule_id,
        severity: p.severity,
        taxonomy: p.taxonomy,
        message: `Description poisoning heuristic matched: ${p.rule_id}`,
        evidence: String(m[0]).slice(0, 120)
      });
    }
  }
  return findings;
}

function scanShadowAndTyposquat(tools, { similarityThreshold = 0.82 } = {}) {
  const findings = [];
  const names = tools.map((t) => String(t.name || '').trim()).filter(Boolean);
  const seen = {};
  for (const n of names) {
    const key = n.toLowerCase();
    seen[key] = (seen[key] || 0) + 1;
  }
  for (const [key, count] of Object.entries(seen)) {
    if (count > 1) {
      findings.push({
        tool_name: key,
        rule_id: 'shadow_tool_duplicate',
        severity: 'High',
        taxonomy: 'MCP03',
        message: `Shadow tool: name appears ${count} times`,
        evidence: key
      });
    }
  }
  for (let i = 0; i < names.length; i += 1) {
    for (let j = i + 1; j < names.length; j += 1) {
      if (names[i].toLowerCase() === names[j].toLowerCase()) continue;
      const score = ratio(names[i], names[j]);
      if (score >= similarityThreshold) {
        findings.push({
          tool_name: names[i],
          rule_id: 'typosquat_lookalike',
          severity: 'High',
          taxonomy: 'MCP08',
          message: `Typosquat / lookalike pair: '${names[i]}' ~ '${names[j]}' (${score.toFixed(2)})`,
          evidence: `${names[i]}|${names[j]}|${score.toFixed(2)}`
        });
      }
    }
  }
  return findings;
}

function scanTools(tools) {
  const findings = [];
  for (const tool of tools) findings.push(...scanDescription(tool));
  findings.push(...scanShadowAndTyposquat(tools));
  const ok = !findings.some((f) => f.severity === 'Critical' || f.severity === 'High');
  return { ok, tools_scanned: tools.length, findings };
}

module.exports = { scanTools, scanDescription, scanShadowAndTyposquat };
